File: hard/2751-robot-collisions/solution.java
import java.util.*;

/**
 * 2751. Robot Collisions
 * Time: O(n log n)
 * Space: O(n)
 */
class Solution {
    public List<Integer> survivedRobotsHealths(int[] positions, int[] healths, String directions) {
        int n = positions.length;

        Integer[] indices = new Integer[n];
        for (int i = 0; i < n; i++) indices[i] = i;

        Arrays.sort(indices, (a, b) -> positions[a] - positions[b]);

        Stack<Integer> stack = new Stack<>();

        for (int idx : indices) {
            if (directions.charAt(idx) == 'R') {
                stack.push(idx);
            } else {
                while (!stack.isEmpty() && healths[idx] > 0) {
                    int top = stack.peek();

                    if (healths[top] < healths[idx]) {
                        healths[idx]--;
                        healths[top] = 0;
                        stack.pop();
                    } else if (healths[top] > healths[idx]) {
                        healths[top]--;
                        healths[idx] = 0;
                        break;
                    } else {
                        healths[top] = 0;
                        healths[idx] = 0;
                        stack.pop();
                        break;
                    }
                }
            }
        }

        List<Integer> result = new ArrayList<>();
        for (int h : healths) {
            if (h > 0) result.add(h);
        }

        return result;
    }
}

File: hard/2751-robot-collisions/solution.py
from typing import List


class Solution:
    def survivedRobotsHealths(self, positions: List[int], healths: List[int], directions: str) -> List[int]:
        """
        2751. Robot Collisions
        Time: O(n log n)
        Space: O(n)
        """
        n = len(positions)

        robots = sorted([(positions[i], i) for i in range(n)])

        stack = []

        for _, i in robots:
            if directions[i] == 'R':
                stack.append(i)
            else:
                while stack and healths[i] > 0:
                    j = stack[-1]

                    if healths[j] < healths[i]:
                        healths[i] -= 1
                        healths[j] = 0
                        stack.pop()
                    elif healths[j] > healths[i]:
                        healths[j] -= 1
                        healths[i] = 0
                        break
                    else:
                        healths[j] = 0
                        healths[i] = 0
                        stack.pop()
                        break

        return [healths[i] for i in range(n) if healths[i] > 0]
